hilti/toolchain/src/ast/scope.cc

```diff
@@ -11,20 +11,21 @@
 
 using namespace hilti;
 
-void Scope::insert(const ID& id, NodeRef&& n) {
-    assert(n && n->isA<Declaration>());
-    const auto& d = n->as<Declaration>();
-    auto& nodes = _items[std::string(id)];
-
-    // Filter out duplicates
-    for ( const auto& i : nodes ) {
-        if ( i && i->as<Declaration>() == d )
-            return;
-    }
+std::size_t Scope::NodeRefHash::operator()(const hilti::NodeRef& n) const {
+    if ( n )
+        return reinterpret_cast<size_t>(&*n);
```

Comment:
So the hash is the address of the node, right? That's changing semantics slightly, in the case that two declarations exist that are semantically equivalent but stored in separates nodes. But I think that's fine, I can't really see how that would become a problem.

Reply:
It turns out that this approach does not work reliably and by using the address we might end up hashing node refs which should be considered equal to different buckets.

I adjusted this to instead use a) the presence of a node behind the ref, and b) a hash of the location information. This seems to work.